### PR TITLE
Fix component data in Gradle Module Metadata

### LINKED ISSUE #19769: Wrong Gradle module metadata json when multiple publications share the same component
We have a project that publishes artifacts to multiple Maven coordinates with different publications (from the same component). In our case, we publish either release or snapshot versions, depending on which Gradle task we run:

```
publishing {
    publications {
        release(MavenPublication) {
            groupId = 'com.example'
            artifactId = 'test-library'
            version = '1.0'
            from components.java
        }
        snapshot(MavenPublication) {
            groupId = 'com.example'
            artifactId = 'test-library'
            version = '1.0-SNAPSHOT'
            from components.java
        }
    }
}
```

Gradle also pushes a ".module" file to Maven for gradle-specific module metadata. Upon inspecting these files we found that the coordinates listed in there are not correct.

### Expected Behavior
The generated module metadata has the correct maven coordinates from the publication. Specifically, for "release" we expect:
```
{
  "formatVersion": "1.1",
  "component": {
    "group": "com.example",
    "module": "test-library",
    "version": "1.0",
    "attributes": {
      "org.gradle.status": "release"
    }
  },
  ...
```

### Current Behavior
The generated module metadata has the maven coordinates from the alphabetically last publication with the same component, in our case from "snapshot".
```
{
  "formatVersion": "1.1",
  "component": {
    "group": "com.example",
    "module": "test-library",
    "version": "1.0-SNAPSHOT",
    "attributes": {
      "org.gradle.status": "integration"
    }
  },
  ...
```

### Context
We want to publish the same artifact to different coordinates (which are tied to different physical repositories) without having to edit the .gradle files every time we wish to push a release, and have been using the above approach to accomplish that. It worked fine until we recently moved from Gradle 5.6.x to 7 (I know, long overdue maintenance) in which the maven publish plugin publishes .module files. If there's a better way to achieve the same goal, we are happy to implement that.

For now we disabled metadata generation to work around the issue.

### Steps to Reproduce 
See reproducer: https://github.com/bvschaik/gradle-wrong-module-metadata-reproducer

1. Add two publications to `build.gradle`:
```
plugins {
    id 'java'
    id 'maven-publish'
}

repositories {
    mavenCentral()
}

publishing {
    publications {
        release(MavenPublication) {
            groupId = 'com.example'
            artifactId = 'test-library'
            version = '1.0'
            from components.java
        }
        snapshot(MavenPublication) {
            groupId = 'com.example'
            artifactId = 'test-library'
            version = '1.0-SNAPSHOT'
            from components.java
        }
    }
}
```

2. Run the `generateMetadataFileFor[..]Publication` tasks.
3. Inspect the generated module files

### Your Environment
Building locally on MacOS, building on CI on Linux. Environment doesn't influence reproducibility. I tried Gradle 7.0.2, 7.3.3, and locally built from latest source (7.5) which all have the issue.

### Additional info
I looked into the source of Gradle:

- `GenerateModuleMetadata.computeModuleMetadataSpec()` creates a ModuleMetadataSpec using the builder
- `ModuleMetadataSpecBuilder()` constructor calls `collectCoordinates()` which creates a map from the publications: `publication.component` to `ComponentData` which is later used to get the maven coordinates for the metadata file
- Since both publications use the same component (`java`), the last one will overwrite the earlier's data

### PR DESCRIPTION
**Pull request by jbartok**
Fixes #19769

### CODE REVIEW

**Comment by jbartok** [comment]
@bot-gradle test RFM

**Comment by jbartok** [comment]
@bot-gradle

**Comment by jbartok** [comment]
@bot-gradle test RFN

File: subprojects/publish/src/test/groovy/org/gradle/api/publish/internal/metadata/GradleModuleMetadataWriterTest.groovy [hunk 1]
```diff
@@ -1160,6 +1160,90 @@ class GradleModuleMetadataWriterTest extends Specification {
 """
     }
 
+    // fixing a problem reported in https://github.com/gradle/gradle/issues/19769
```
Comment (ljacomet) [inline subprojects/publish/src/test/groovy/org/gradle/api/publish/internal/metadata/GradleModuleMetadataWriterTest.groovy:1163]: Use `@Issue` for this, there is an example in the same file line 878

**Review by ljacomet — CHANGES_REQUESTED** [review]
LGTM!

Only a minor change to reference the issue:

File: subprojects/publish/src/test/groovy/org/gradle/api/publish/internal/metadata/GradleModuleMetadataWriterTest.groovy (hunk 1 above)
Comment (jbartok) [reply subprojects/publish/src/test/groovy/org/gradle/api/publish/internal/metadata/GradleModuleMetadataWriterTest.groovy]: Done

**Review by ljacomet — APPROVED** [review]

**Pull request merged**